### DISCUSSION TITLE
Opaque type error for custom error type in `pytest.raises`

### DISCUSSION THREAD

**Discussion by dan-ri** [microsoft/pyright#3988] [Q&A] [answered]
Python 3.8.3, Pytest 6.2.5, Pyright 1.1.272:

Seeing widespread occurrences of this kind of error for custom error types in a codebase:

```python
with pytest.raises(CustomError):
    ...
```
```
test_file.py:1000:9 - error: Argument of type "Type[Type[CustomError]]" cannot be assigned
to parameter "expected_exception" of type "Type[_E@raises] | Tuple[Type[_E@raises], ...]" in function "raises"
    Type "Type[Type[CustomError]]" cannot be assigned to type "Type[_E@raises] | Tuple[Type[_E@raises], ...]"
      Type "Type[Type[CustomError]]" cannot be assigned to type "Type[_E@raises]"
      "Type[type]" is incompatible with "Type[Tuple[Type[_E@raises], ...]]" (reportGeneralTypeIssues)
```
`CustomError` is a class ultimately derived from `Exception`, with a few layers in between. I assume this means there's something about `CustomError` that makes pyright think it's not suitable as an argument to `pytest.raises`, but it's awfully hard to tell what that might be from this diagnostic. Any ideas?

**Comment by erictraut** [comment, marked as answer]
The latest version of pytest is a "py.typed" library, which means that it ships with inlined type information. Please confirm that you're using the latest version of pytest.

To diagnose the problem, I'll need a repro case — preferably a self-contained code sample. Your code above references a symbol `CustomError` which is not defined. 

Based on the error message, I'm guessing that `CustomError` does not derive from `BaseException`, which is required by the `pytest.raises` function, at least according to its type declaration.

**Reply by dan-ri** [reply]
Thanks for the quick response! This is with Pytest 6.2.5, so not the absolute latest but fairly recent.

`CustomError` ultimately derives from `Exception` (hence `BaseException`) but there are a few classes in between, e.g.

```python
class CustomError(OtherErrorA)
    ...
class OtherErrorA(OtherErrorB)
    ...
class OtherErrorB(OtherErrorC)
    ...
class OtherErrorC(Exception)
    ...
```

I will try to dig into this more and provide a minimal reproduction.

**Reply by dan-ri** [reply]
Here's a minimal case of what I'm seeing:

```python
from typing import Callable, Type, TypeVar
import pytest

class CustomErrorBase(Exception):
    pass

_T = TypeVar("_T", bound=Type[CustomErrorBase])

def error_decorator() -> Callable[[_T], Type[_T]]:
    def decorator(error: _T) -> Type[_T]:
        class GeneratedError(error):  # type: ignore
            pass
        return GeneratedError
    return decorator

@error_decorator()
class CustomError(CustomErrorBase):
    pass

def oops() -> None:
    with pytest.raises(CustomError):
        pass
```

This passes with mypy due to the `# type: ignore`, but pyright returns the error above.

**Reply by JelleZijlstra** [reply]
The annotation on `error_decorator` is wrong. It should return `_T`, not `Type[_T]`.

**Reply by erictraut** [reply]
Yes, @JelleZijlstra is correct in his diagnosis, but I think there's a better fix.

I would generally not recommend using a `Type[X]` as the upper bound for a `TypeVar`. Instead, I would do the following:

```python
_T = TypeVar("_T", bound=CustomErrorBase)

def error_decorator() -> Callable[[Type[_T]], Type[_T]]:
    def decorator(error: Type[_T]) -> Type[_T]:
        class GeneratedError(error):  # type: ignore
            pass

        return GeneratedError

    return decorator
```

**Reply by dan-ri** [reply]
Thank you @erictraut and @JelleZijlstra! After applying the suggested type fix, the `pytest.raises` calls typecheck successfully in both mypy and pyright.
